### DISCUSSION TITLE
struct bit fields support?

### DISCUSSION THREAD

**Discussion by liujoey** [vlang/v#11776] [Ideas]
Is there any plan to support bit fields in struct? For some low level programming or binary file format definition, bit fields are required. I was planning to use V for one of my project but the major obstacle is the lack of bit fields feature.

**Comment by medvednikov** [comment]
We already do, it's just not documented yet.

**Comment by liujoey** [comment]
Thank you. Is there a quick intro for the syntax? I'm new to the Vlang so I couldn't find the syntax for bit fields.

**Comment by liujoey** [comment]
OK, I searched around and found BitField module in vlib. From the source I understand that this module is about bit array manipulation. I believe it is different from what I'm trying to talk here. The feature I talked about in this issue is actually a C struct syntax:

```c
#pragma pack(push, 1)
struct AnnotationTableHeader {
    bool two_dimensional_main : 1;          // u(1)
    SymmetryMode symmetry_mode : 3;         // u(3)
    bool symmetry_minor_diagonal : 1;       // u(1)
    uint8_t n_aux_attribute_groups : 3;     // u(3)
    uint8_t ann_table_index_size : 2;       // u(2)
    # ... more fields
};
#pragma pack(pop)
```

When I serialize this struct into a binary file or send it over through socket, the first 4 fields would be packed into 1 byte. I'm looking for something like this in V.

For now I will just leave these definition in c code and try to write a V module to wrap it. I would really appreciate if this is directly supported in V syntax.

Appreciate for your help.

**Comment by medvednikov** [comment]
@joe-conigliaro can you please post here an example of a struct bit field you worked on?

Will add it to the docs.

**Comment by Cons-Cat** [comment]
Has this been clarified somewhere else?

**Comment by Gladear** [comment]
Not to my knowledge, but if you need to use it now, check out [this test file](https://github.com/vlang/v/blob/master/vlib/v/tests/enum_bitfield_test.v) as a reference
